### PR TITLE
Research ai

### PR DESCRIPTION
**Pull request by Dilvish-fo**
This is a continuation from freeorion/freeorion#188  I'm establishing this branch on my own fork to facilitate our whole AI team being able to more directly work together on this a fair bit more (rather than being limited to asking luiges90 to take care of all changes) before we merge it into main master.  

There are two main aspects I see to the work that needs to be done-- one aspect is restructuring this a bit to make it easier to work with, and then the other aspect is adjusting it and tuning it to get results we're ready to put into main master.

On the first task, it seems to me that get_priority() winds up a bit bogged down by trying to cover so much in one place, and that also tends to lead to things winding up very separated, like "engines" being defined at one spot but then not being used until ~150 lines later.  I am a bit inclined to have get_priority(tech_name) itself simply call an appropriate helper function like get_def_priority(tech_name) based on the prefix of tech_name.  Then each of these helper functions would be much more tightly focused & easier to think about in a clear manner (I think).

Does anyone else have suggestions/comments on the restructuring aspect?

I'll go ahead accept this initial PR into this branch, but we can continue discussing things here., and any of us can open new PRs into this branch with further contributions, etc.

### CODE REVIEW

File: default/AI/FreeOrionAI.py [hunk 1]
```diff
@@ -166,6 +166,7 @@ def generateOrders():  # pylint: disable=invalid-name
     random_seed = str(fo.getGalaxySetupData().seed) + "%03d%05d" % (fo.empireID(), turn)
     random.seed(random_seed)
     aggression_name = fo.aggression.values[foAIstate.aggression].name
+
```
Comment (Cjkjvfnby) [inline default/AI/FreeOrionAI.py:169]: How about remove this empty line?

Comment (Dilvish-fo) [inline default/AI/FreeOrionAI.py:169]: yes, that's a comment I already made on the other PR, but I didn't want to ask luiges to do further cleanup of history before getting things going here.

**Comment by Morlic-fo** [comment]
> I am a bit inclined to have get_priority(tech_name) itself simply all an appropriate helper function like get_def_priority(tech_name) based on the prefix of tech_name.

Either this or we have some basic calculation for the respective priorities once per turn (at least the dynamic part, static part should be cached) and save the priorities in a dict. I do not really see a reason why we would want to calculate the priorities multiple times for each queried tech.

So we would have something like

```
def get_priority(tech_name):
    if not updated_this_turn:
        calculate_priorities()
    if tech_name in AIDependencies.DEFENSE_TECHS:
        return priorities['defense']
    ...
```

calculate_priorities() then should probably split into multiple functions as you suggested.

**Comment by Dilvish-fo** [comment]
@luiges90 Do you want to take a first stab at this refactoring?

**Comment by luiges90** [comment]
@Dilvish-fo sure, plan to update priority functions along with refactoring

**Comment by Dilvish-fo** [comment]
@luiges90 -- it sort of looks like you've finished all or most of a first pass at the refactoring, and have started moving on to substantive details, is that right?

It seems like it's probably time for us to do a bit of review and discussion on that refactoring stage.  As part of that, it might be good to also do some of the cleanup/squashing that Marcel talks about here: http://www.freeorion.org/forum/viewtopic.php?p=76662#p76662

As an example, I went ahead and did that with your current progress, mostly very simply just combining adjacent commits, leaving out messages about fixing a bug in a previous part of the same group, since once they're squashed together the bug just no longer exists.  The one thing that I broke out as a fragment was the addition of tags to species.txt, just since that seemed a bit distinctive (and I wanted a separate commit on it to maybe help remind content scripters to try to keep those up to date themselves).  I added a little note at the end of the intro line for each commit message indicating which of your underlying commits it corresponded to.

So, my suggestion is in a new branch; I think this PR here, will not be so useful for more discussions going forward, so I opened a new one from this new cleanup branch, and suggest we do review/discussions on the commits in that branch.  https://github.com/Dilvish-fo/freeorion/pull/5

We could go with those commit groupings or you redo your own if you think some different ones make more sense.

**Comment by Dilvish-fo** [comment]
opps, I just realized I hadn't checked the boards this morning and then saw your post there: http://www.freeorion.org/forum/viewtopic.php?p=79399#p79399

So we can do the discussion at that location.  Please take a glance at the groupings I did though, I think some history cleanup like that will be best before putting it into master.

**Pull request merged**